torch/_decomp/decompositions.py

```diff
@@ -3282,66 +3333,68 @@ def _compute_source_index(scale, dst_index, align_corners):
-        input, type_promotion_kind=utils.ELEMENTWISE_TYPE_PROMOTION_KIND.INT_TO_FLOAT
-    )
-    # We have to create arange with int64 dtype and use .to in order to avoid
-    # additional kernels creation in inductor and get a perf slowdown
-    i = torch.arange(output_size[0], device=input.device).to(dtype=dtype)
-    j = torch.arange(output_size[1], device=input.device).to(dtype=dtype)
-
-    x_f32 = _compute_source_index(w_scale_factor, j, align_corners).clamp(min=0.0)
-    y_f32 = _compute_source_index(h_scale_factor, i, align_corners).clamp(min=0.0)
-    y_f32 = y_f32.unsqueeze(-1)
-
-    x = x_f32.to(torch.int64)
-    y = y_f32.to(torch.int64)
-
-    # We are using torch.where instead of torch.clamp below due to an expected failure
-    # in test_aot_autograd_symbolic_exhaustive_nn_functional_interpolate_bilinear_cpu_float32 test
-    # torch.ops.aten.clamp.default(add, None, sub) on int64 input tensor is returning float32 and
-    # fails with torch.ops.aten._unsafe_index.Tensor(primals_1, [None, None, _to_copy_1, clamp_2])
-    # RuntimeError: _unsafe_index found unexpected index type Float
-    # xp1 = (x + 1).clamp(max=in_w - 1); yp1 = (y + 1).clamp(max=in_h - 1)
-    xp1 = torch.where(x < in_w - 1, x + 1, x)
-    yp1 = torch.where(y < in_h - 1, y + 1, y)
-
-    v1 = aten._unsafe_index(input, [None, None, y, x])
-    v2 = aten._unsafe_index(input, [None, None, y, xp1])
-    v3 = aten._unsafe_index(input, [None, None, yp1, x])
-    v4 = aten._unsafe_index(input, [None, None, yp1, xp1])
+    n_batch, n_channels = input.shape[:2]
+    inp_size = input.shape[2:]
+    d = len(inp_size)
+
+    def get_values(inp, out, scales, nsqueeze):
+        # First Calculate scaling factor
+        scale_factor = _compute_scale(inp, out, align_corners, scales)
+        _, dtype = utils.elementwise_dtypes(
+            input,
+            type_promotion_kind=utils.ELEMENTWISE_TYPE_PROMOTION_KIND.INT_TO_FLOAT,
+        )
+        # We have to create arange with int64 dtype and use .to in order to avoid
+        # additional kernels creation in inductor and get a perf slowdown
+        i = torch.arange(out, device=input.device).to(dtype=dtype)
+
+        x_f32 = _compute_source_index(scale_factor, i, align_corners).clamp(min=0.0)
+        x_f32 = x_f32.reshape(len(x_f32), *[1] * (nsqueeze))
+        x = x_f32.to(torch.int64)
+
+        # We are using torch.where instead of torch.clamp below due to an expected failure
+        # in test_aot_autograd_symbolic_exhaustive_nn_functional_interpolate_bilinear_cpu_float32 test
+        # torch.ops.aten.clamp.default(add, None, sub) on int64 input tensor is returning float32 and
+        # fails with torch.ops.aten._unsafe_index.Tensor(primals_1, [None, None, _to_copy_1, clamp_2])
+        # RuntimeError: _unsafe_index found unexpected index type Float
+        # xp1 = (x + 1).clamp(max=in_w - 1); yp1 = (y + 1).clamp(max=in_h - 1)
+        xp1 = torch.where(x < inp - 1, x + 1, x)
```

Comment:
The failures on the SAM models look real. I think they may be coming from this code. See if doing
```suggestion
        xp1 = torch.where(x < inp - 1, x + 1, inp - 1)
```
fixes them. I'd hope this not to change much the generated code, but we should check that

Reply:
With the latest fix, the SAM model failure is gone too.